README
Script:			SeVeNblend
Author:			Piotr Nowicki
Date:			2008-04-29
Last Change:		2009-10-06
======================================
 
I. INTRODUCTION

This script was written to make the whole process of compiling Blender from SVN with Yaf(a)ray (rewritten core of Yafray) easier (setting paths, adding volume_integrator, copying, etc.).


II. WHAT DOES THIS SCRIPT DO

1. Download Blender, Yaf(a)ray Core and Yaf(a)ray Exporter from SVN
2. Compile it
3. Let you run it and have a fun :)

Also you can:
1. Compile Blender with your own icon theme
2. Optimize it for your CPU


III. REQUIRED LIBRARIES

The script needs few libraries in order to work properly (notice than if you edit user-config.py you might have to install some more libraries) like:
- subversion
- scons
- png
- gcc
- g++
- alut
- sdl
- openexr
- jpeg
- python (2.5 would be nice)
- tiff
- gettext
- xi
- freetype
- xml2
- ... and everything that depends on those

For Ubuntu users - you could try to copy and paste this line (or adjust it if mismatches occurs):
sudo apt-get install subversion libqt4dev binutils swig scons libpng12-dev gcc g++ libalut-dev libsdl-sound1.2-dev libopenexr-dev libjpeg62-dev python2.5-dev libtiff4-dev gettext libxi-dev libfreetype6-dev libxml2-dev


IV. USAGE

./sevenblend.sh [OPTIONS...]

Where OPTIONS are :
!!! THIS SECTION IS OBSOLETE. PLEASE TYPE ./sevenblend HELP for full options list with description!!!:
- WITH_BUTT - yea, kind of funny switch ;) Compile blender with your own buttons (you need to put your icons into 'addons' directory of the script and name it 'blenderbuttons.png'. By default it's compiled with Blender standard buttons.

- WITH_OPTIM - use user-config.py to optimize build. The 'user-config.py' script must be in 'user_configs' directory of the script. There is one user-config.py which was used by me for Intel Core2Duo build on gcc 4.2 and another, very short, config for AMD64 processors. By default it's compiled with default user-config.py.

- FORCE - When there are no changes in any of three repositories (Blender and 2x Yaf(a)ray) there is no need of recompiling Blender. BUT if you want to add new buttons or try new optimimalization configuration - use this switch to tell script to fore recompilation of Blender source.

- HELP - shows what switches you can use with the script (basicaly short version of this section

- NO_YAFARAY - compile only Blender - without inserting Yaf(a)ray into Blender source; by default standard Yafray will be there instead


V. INSTRUCTIONS

1. Unpack the script (if you're reading it you must have done it already ;)
2. Be sure that the unpacked directory (which contains sevenblend.sh file) has rights to write/read/execute (chmod 700)
2. Change permissions to allow owner to execute sevenblend (ex. chmod 500 ./sevenblend.sh)
3. Run it with or without any of the swiches mentioned in IV.
4. Read the informations on screen and be patient.
5. When prompted about the permission to access sites via SVN accept temporary or permanently
6. There are two sudo's commands in script - one is for installing of yaf(a)ray into /usr/local/ and another is for running ldconfig
7. After installation there will be made a shortcut for blender build in your home directory named 'sevenblend' which you might want to run and test your new Blender :)
8. When running Blender, set 'scripts path' in Blender options to point to 'YOUR_SEVENBLEND_DIR/.sources/install/linux2/.blender/scripts' directory in order to use Yaf(a)ray export scripts

VI. FILE STRUCTURE

.
|-- README				- the file you're reading right now
|-- addons
|   `-- blenderbuttons.png		- modified icons to use
|-- sevenblend.sh			- main 'heart' of the script 
|-- user_configs
|   |-- AMD64_user-config.py		- example config (rename to user-config.py)
|   |-- CORE2DUO_user-config.py		- example config (rename to user-config.py)
|   `-- user-config.py			- file to use when WITH_OPTIM switch set
`-- yafaray_test.blend			- example scene to render with yaf(a)ray

VII. BUNDLE

Latest bundle of SeVenBlend is 0.2.2 (10.2009) and can be downloaded from http://www.piotrnowicki.com/sevenblend/sevenblend-0.2.2.tar.gz.

VIII. CONTACT

If you would like to contact me, please feel free to write at pallipp@gmail.com with any problems with this script. If I can - I will gladly help.

Have fun!


CHANGE_LOG
2009-10-07 Palli
* Version 0.2.2.
	- Changed repositories of Yafaray Core + Exporter to the new ones.

2009-10-05 Palli
* Version 0.2.1.
	- Added fix for 'could not find nasm.inc' bug,
	- Added switch to select if you want to compile Blender from trunk (v. 2.5) or 2.4 branch,
	- Added some packages to INSTALL_PCKGS switch,

2008-12-26 Palli
* Version 0.2.0.
	- Removed 0.0.2 export support; only 0.0.3 with QT render is now available
	- INSTALL_PCKGS switch. Installs needed packages for compilation process. 
	    Uses apt-get tool so it's mainly usable for Debian family distros
	- NO_QT switch. Do not build QT GUI for Yafaray (runs with 0.0.3 exporter script)
	- CLEAN switch. Cleans build directories and exits
	- CLEAN_ALL switch. Cleans all directories (src + build + logs) and exits
	- New Blender buttons pack added (jendrzych, http://blenderartists.org/forum/showthread.php?t=84971)
	- NO_YAFARAY switch has been removed
	- Scons multiple targets parameter by default set to '2' instead of '1' 

2008-12-22 Palli
* Version 0.1.4.
	- Added Yaf(a)ray Exporter script ver. 0.0.3 Python script 
	    (it's available next to 0.0.2 ver. from Blender scripts menu so you can choose the one you prefer)
	- USE_INTERNAL_SCONS flag. It is up to user to decide if one want to use system scons or the one 
	    that comes with Blender source (default = uses system scons command) 
	    [http://feeblemind.tuxfamily.org/dotclear/index.php/2008/11/02/108-compiling-blender-important-new-upgrade]
	- THREADS_NUM param. Adds '-j' option for scons to compile multiple targets parrarel (default = 1)
	- LINES_TO_DUMP param. Dumps last 'x' lines from log in case of fail (default = 15)
	- REV_BLENDER, REV_YAF_CORE, REV_YAF_EXP params. Specifiec what revision do you want to checkout 
	    of Blender, Yaf(a)ray Core and Yaf(a)ray Exporter (default = 0; it means that the latest revision
	    will be checked out)
	- Clearer information about compilation error and steps (along with time) taken by the script 
	    for easier recognition of problem
	- Added 'swig' and 'binutils' to Ubuntu users apt-get command in README file
	- Added information about setting proper script path in Blender after installing SeVeNBlend
	- Separate directory for all Python scripts created/copied by SeVeNblend

2008-10-04 Palli
* Version 0.1.3.
	- Fixed compiling under GCC 4.3.x where some libraries were missing 
	(thanks Lempour@yafray.org/forum for mentioning them)
	- WITH_BF_REDCODE flag is set to 'false' in user-config.py to prevent
	occuring of Blender building errors
	- BF_QUIET=0 BF_FANCY=0 flags added to scons command. Now one is able
	to visually check logs more clearly than without those flags
	- Fixed checking if volume integrator for preview materials is added 
	to exporter by authors

2008-05-18 Palli
* Version 0.0.1.
   - FORCE switch forces recompilation of Yaf(a)ray [until now it was
   recompiling the Blender only],
   - Proper information about using user-config.py ([INFO] instead of [WARN])

2008-05-05 Palli
* Changes to alpha version.
   - added switches (HELP, NO_YAFARAY),
   - works with changed Blender code where
       BLI_convertstringcode(char*, char*, int) 
	      was replaced by
       BLI_convertstringcode(char*, char*) 


include/arguments.sh
for ARGUMENT in "$@"
do
	if [ $ARGUMENT = "HELP" ]; then
    	print_info "Usage: $0 [OPTION]'"
    	print_info "-- INSTALL_PCKGS - installs packages used by Blender and Yafaray core + exporter compilation (uses apt-get tool)"
    	print_info "-- NO_QT - do not build Yafaray QT GUI"
    	print_info "-- WITH_OPTIM - use compilation optimalization using user-script.py scripts"
		print_info "-- WITH_BUTT - put user-created icons as default ones"
		print_info "-- FORCE - force recompilation even if there is no need (code is up to date)"
		print_info "-- REV_BLENDER=[num] - checkout Blender revision [num] (default = latest)"
		print_info "-- REV_YAF_CORE=[num] - checkout Yaf(a)ray core revision [num] (default = latest)"
		print_info "-- REV_YAF_EXP=[num] - checkout Yaf(a)ray exporter revision [num] (default = latest)"
		print_info "-- THREADS_NUM=[num] - use [num] threads when building with scons (default = 2)"
		print_info "-- LINES_TO_DUMP=[num] - dump last [num] lines from log file when compilation fails (default = 15)"
		print_info "-- USE_INTERNAL_SCONS - use internal (Blender's) scons instead of system's"
		print_info "-- BLENDER2.4 - compile Blender 2.4 line (removed from trunk)"
		print_info "-- CLEAN - delete all compiled resources and quits"
		print_info "-- CLEAN_ALL - delete everything (sources + build + logs) and quits"
		print_info "-- HELP - shows this information"

		echo
		print_info "Example 1.: '$0 WITH_OPTIM FORCE' will force to recompile Blender using optimalization script"
		print_info "Example 2.: '$0 INSTALL_PCKGS LINES_TO_DUMP=25 USE_INTERNAL_SCONS' will download needed packages using apt-get and then build Blender using internal scons from Blender repository and in case of compilation error it will list last 25 lines on the screen."
		echo
		
		# We've told how the script works - so we can safely exit ;)
		exit 1
	
	#######
	# SWITCH - DO NOT INSTALL YAFARAY QT GUI
    #
    elif [ $ARGUMENT = "NO_QT" ]; then
    	print_info "Not building Yafaray QT GUI."
     	NO_QT_GUI=1
  
 	#######
	# SWITCH - INSTALL PACKAGES NEEDED FOR COMPILATION USING APT-GET
    #
    elif [ $ARGUMENT = "INSTALL_PCKGS" ]; then
    	print_info "Installing needed packages."
     	INSTALL_PACKAGES=1
     	    
     	        	    
	#######
	# SWITCH - OPTIMALIZATION
    #
    elif [ $ARGUMENT = "WITH_OPTIM" ]; then
    	print_info "Optimalization should be turned ON: using '$USER_CONFIG_SCRIPT'"
    	if [ -f "$USER_CONFIG_SCRIPT" ]; then
    		print_info "-- File '$USER_CONFIG_SCRIPT' found! Using it." 
    		SHOULD_OPTIMIZE=1
    	else
    		print_warn "-- File '$USER_CONFIG_SCRIPT' NOT found! Turning optimalization OFF!" 
    		SHOULD_OPTIMIZE=0
    	fi
    	
	#######
	# SWITCH - BUTTONS
    #
    elif [ $ARGUMENT = "WITH_BUTT" ]; then
    	print_info "Modified icons will be included into Blender build."
    	if [ -f "$ADDONS/blenderbuttons.png" ]; then
    		print_info "-- File '$ADDONS/blenderbuttons.png' found! Using it!" 
    		SHOULD_INSTALL_ICONS=1
    		#SHOULD_COMPILE_BLENDER=1
    	else
    		print_warn "-- File '$ADDONS/blenderbuttons.png' NOT found! Modified icons not included!" 
    		SHOULD_INSTALL_ICONS=0
    	fi
    	
    #######
	# SWITCH - FORCE RECOMPILATION
    #
  	elif [ $ARGUMENT = "FORCE" ]; then
    	print_info "Forcing recompiling of Blender and Yaf(a)ray Core."
     	SHOULD_COMPILE_BLENDER=1
		SHOULD_COMPILE_YAFARAY=1

	#######
	# SWITCH - USE INTERNAL BLENDER SCONS
    #
    elif [ $ARGUMENT = "USE_INTERNAL_SCONS" ]; then
    	print_info "Using internal Blender scons."
     	USE_INTERNAL_SCONS=1
     	
	#######
	# SWITCH - CLEAN BUILD DIRECTORIES
    #
    elif [ $ARGUMENT = "CLEAN" ]; then
    	print_info "Will clean build directories."
     	CLEAN_BUILD=1
     	    
	#######
	# SWITCH - CLEAN SRC+BUILD DIRECTORIES
    #
    elif [ $ARGUMENT = "CLEAN_ALL" ]; then
    	print_info "Will clean all files."
     	CLEAN_ALL=1    
     	      	         	     	
	#######
	# VALUE - BLENDER REVISION
    #
    elif [ `echo "$ARGUMENT"|cut -f 1 -d '='` = "REV_BLENDER" ]; then
    	print_info "Specified Blender Revision."
    	REV_BLENDER=`echo "$ARGUMENT"|cut -f 2 -d '='`

		if echo "$REV_BLENDER"|cut -f 2 -d '='|grep -qE ^[0-9]+$; then 
			print_info "-- Revision number $REV_BLENDER will be compiled."
		else
			print_warn " Revision number '$REV_BLENDER' is in wrong format (only digits are accepted). Newest revision will be compiled."
			REV_BLENDER=0
		fi
     	
   
   	#######
	# VALUE - YAFARAY CORE REVISION
    #
    elif [ `echo "$ARGUMENT"|cut -f 1 -d '='` = "REV_YAF_CORE" ]; then
    	print_info "Specified Yaf(a)ray Core Revision."
    	REV_YAF_CORE_TMP=`echo "$ARGUMENT"|cut -f 2 -d '='`

		if is_digit "$REV_YAF_CORE_TMP"; then 
			print_info "-- Revision number $REV_YAF_CORE_TMP will be compiled."
			REV_YAF_CORE=$REV_YAF_CORE_TMP
		else
			print_warn " Revision number '$REV_YAF_CORE_TMP' is in wrong format (only digits are accepted). Newest revision will be compiled."
		fi
		
   	#######
	# VALUE - YAFARAY EXPORTER REVISION
    #
	elif [ `echo "$ARGUMENT"|cut -f 1 -d '='` = "REV_YAF_EXP" ]; then
    	print_info "Specified Yaf(a)ray Exporter Revision."
    	REV_YAF_EXP_TMP=`echo "$ARGUMENT"|cut -f 2 -d '='`

		if is_digit "$REV_YAF_EXP_TMP"; then 
			print_info "-- Revision number $REV_YAF_EXP_TMP will be compiled."
			REV_YAF_EXP=$REV_YAF_EXP_TMP
		else
			print_warn " Revision number '$REV_YAF_EXP_TMP' is in wrong format (only digits are accepted). Newest revision will be compiled."
		fi
		
	#######
	# VALUE - SCONS THREADS NUMBER
    #
	elif [ `echo "$ARGUMENT"|cut -f 1 -d '='` = "THREADS_NUM" ]; then
    	print_info "Specified Scons parrarel target to be built."
    	THREADS_NUM_TMP=`echo "$ARGUMENT"|cut -f 2 -d '='`

		if is_digit "$THREADS_NUM_TMP"; then 
			print_info "-- $THREADS_NUM_TMP targets will be built parrarel."
			THREADS_NUM=$THREADS_NUM_TMP
		else
			print_warn " Number '$THREADS_NUM_TMP' is in wrong format (only digits are accepted). 2 targets will be build simultaneously."
		fi
 
	#######
	# VALUE - LINES OF LOGS TO BE DUMPED
    #
	elif [ `echo "$ARGUMENT"|cut -f 1 -d '='` = "LINES_TO_DUMP" ]; then
    	print_info "Specified lines of logs to be dumped."
    	LINES_TO_DUMP_TMP=`echo "$ARGUMENT"|cut -f 2 -d '='`

		if is_digit "$LINES_TO_DUMP_TMP"; then 
			print_info "-- Last $LINES_TO_DUMP_TMP will be dumped."
			LINES_TO_DUMP=$LINES_TO_DUMP_TMP
		else
			print_warn " Number '$LINES_TO_DUMP_TMP' is in wrong format (only digits are accepted). Default numer for dumped line is 15."
		fi
		
	#######
	# SWITCH - SHOULD WE COMPILE BLENDER 2.4 (WHICH IS NOT IN TRUNK ANYMORE)
    #
	elif [ `echo "$ARGUMENT"|cut -f 1 -d '='` = "BLENDER2.4" ]; then
    	print_info "Compiling Blender 2.4 line. This will not be the latest version of Blender!"
    	COMPILE_BLENDER_2_4=1 	         	  	     	    	     
     	       	     	
    #######
	# SWITCH - UNKNOWN
    #
    else
    	print_info "Unknown argument '$ARGUMENT'. Ommiting."
    fi
done